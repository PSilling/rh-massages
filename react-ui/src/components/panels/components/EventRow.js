// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import moment from "moment";

// component imports
import MassageEventModal from "../../modals/MassageEventModal";

// util imports
import _t from "../../../util/Translations";
import Auth from "../../../util/Auth";
import Util from "../../../util/Util";

/**
 * Custom component event for MyMassages agenda calendar view.
 */
class EventRow extends Component {
  state = { active: false };

  handleToggle = () => {
    this.setState(prevState => ({ active: !prevState.active }));
  };

  handleKeyPress = event => {
    if (event.key === "Enter") {
      this.handleToggle();
    }
  };

  getUserInfo = () => {
    if (Auth.isMasseur()) {
      if (this.props.event.massage.client === null) {
        return <span style={{ color: Util.SUCCESS_COLOR }}>{_t.translate("Free")}</span>;
      }
      return (
        <span style={{ color: Util.ERROR_COLOR }}>
          {`${this.props.event.massage.client.name} ${this.props.event.massage.client.surname}`}
        </span>
      );
    }

    return <span>{`${this.props.event.massage.masseuse.name} ${this.props.event.massage.masseuse.surname}`}</span>;
  };

  render() {
    const dateString = moment(this.props.event.massage.date).format("L");
    return (
      <tr>
        <td>{this.props.event.massage.facility.name}</td>
        <td>{dateString.substring(0, dateString.length - 5)}</td>
        <td>
          {`${moment(this.props.event.massage.date).format("LT")} â€“ ${moment(this.props.event.massage.ending).format(
            "LT"
          )}`}
        </td>
        <td colSpan="2">
          <strong
            onClick={this.handleToggle}
            role="button"
            style={{ cursor: "pointer" }}
            tabIndex="-1"
            onKeyPress={this.handleKeyPress}
          >
            {`${_t.translate("Massage")}: `}
            {this.getUserInfo()}
          </strong>
          <br />

          {this.state.active && (
            <MassageEventModal
              event={this.props.event}
              label={Auth.isMasseur() ? _t.translate("Force cancel") : _t.translate("Unassign me")}
              allowEditation={false}
              allowDeletion={false}
              onClose={this.handleToggle}
              onConfirm={() => {
                this.handleToggle();
                this.props.onCancel(this.props.event.massage);
              }}
            />
          )}
        </td>
      </tr>
    );
  }
}

EventRow.propTypes = {
  /** event to be rendered */
  event: PropTypes.shape({
    bgColor: PropTypes.string,
    massage: PropTypes.shape({
      id: PropTypes.number,
      masseuse: PropTypes.shape({
        email: PropTypes.string,
        masseur: PropTypes.bool,
        name: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool,
        surname: PropTypes.string
      }),
      date: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      ending: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      client: PropTypes.shape({
        email: PropTypes.string,
        name: PropTypes.string,
        surname: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool
      }),
      facility: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string
      })
    })
  }).isRequired,
  /** function called on massage cancellation */
  onCancel: PropTypes.func.isRequired
};

export default EventRow;
