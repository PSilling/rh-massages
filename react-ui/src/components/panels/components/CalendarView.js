// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import moment from "moment";
import { Table } from "reactstrap";

// component imports
import EventRow from "./EventRow";
import PrintModal from "../../modals/PrintModal";

// util imports
import _t from "../../../util/Translations";
import Util from "../../../util/Util";

class CalendarView extends Component {
  static title() {
    return "";
  }

  static navigate() {
    return new Date();
  }

  state = { printMassages: null };

  setPrintMassages = massages => {
    this.setState({ printMassages: massages });
  };

  createPrintRows = () => {
    const rows = [];
    if (this.state.printMassages.length === 0) {
      rows.push(
        <tr key="info">
          <td colSpan="4">{_t.translate("None")}</td>
        </tr>
      );
    } else {
      for (let i = 0; i < this.state.printMassages.length; i++) {
        rows.push(
          <tr key={i}>
            <td>{moment(this.state.printMassages[i].date).format("L")}</td>
            <td>
              {`${moment(this.state.printMassages[i].date).format("H:mm")}â€“${moment(
                this.state.printMassages[i].ending
              ).format("H:mm")}`}
            </td>
            <td>{`${this.state.printMassages[i].masseuse.name} ${this.state.printMassages[i].masseuse.surname}`}</td>
            <td>
              {Util.isEmpty(this.state.printMassages[i].client)
                ? _t.translate("Free")
                : Util.getContactInfo(this.state.printMassages[i].client)}
            </td>
          </tr>
        );
      }
    }
    return rows;
  };

  render() {
    return (
      <span>
        <div className="no-print">
          <Table hover responsive striped size="sm">
            <thead>
              <tr>
                <th scope="col">{this.props.localizer.messages.date}</th>
                <th scope="col">{this.props.localizer.messages.time}</th>
                <th scope="col">{this.props.localizer.messages.event}</th>
                <th scope="col">
                  <PrintModal events={this.props.events} onPrint={this.setPrintMassages} />
                </th>
              </tr>
            </thead>
            <tbody>
              {this.props.events.map(item => (
                <EventRow key={item.massage.id} event={item} onCancel={this.props.onCancel} />
              ))}
            </tbody>
          </Table>
        </div>

        {this.state.printMassages !== null && (
          <div className="print-only">
            <Table>
              <thead>
                <tr>
                  <th scope="col">{_t.translate("Date")}</th>
                  <th scope="col">{_t.translate("Time")}</th>
                  <th scope="col">{_t.translate("Masseur/Masseuse")}</th>
                  <th scope="col" width="40%">
                    {_t.translate("Client")}
                  </th>
                </tr>
              </thead>
              <tbody>{this.createPrintRows()}</tbody>
            </Table>
          </div>
        )}
      </span>
    );
  }
}

CalendarView.propTypes = {
  /** events featured in the calendar */
  events: PropTypes.arrayOf(
    PropTypes.shape({
      bgColor: PropTypes.string,
      massage: PropTypes.shape({
        id: PropTypes.number,
        masseuse: PropTypes.shape({
          email: PropTypes.string,
          masseur: PropTypes.bool,
          name: PropTypes.string,
          sub: PropTypes.string,
          subscribed: PropTypes.bool,
          surname: PropTypes.string
        }),
        date: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
        ending: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
        client: PropTypes.shape({
          email: PropTypes.string,
          name: PropTypes.string,
          surname: PropTypes.string,
          sub: PropTypes.string,
          subscribed: PropTypes.bool
        }),
        facility: PropTypes.shape({
          id: PropTypes.number,
          name: PropTypes.string
        })
      })
    })
  ).isRequired,
  /** defined localizer translation strings */
  localizer: PropTypes.shape({
    messages: PropTypes.shape({
      date: PropTypes.string,
      time: PropTypes.string,
      event: PropTypes.string
    })
  }).isRequired,
  /** function called on massage cancellation */
  onCancel: PropTypes.func.isRequired
};

export default CalendarView;
