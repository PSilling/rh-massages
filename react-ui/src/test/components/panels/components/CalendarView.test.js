// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import CalendarView from "../../../../components/panels/components/CalendarView";
import EventRow from "../../../../components/panels/components/EventRow";

test("renders content with correct props", () => {
  const testFunction = jest.fn();
  const testEvents = [
    {
      massage: {
        id: 1,
        date: new Date(0),
        ending: new Date(1000),
        masseuse: {
          sub: "m-sub",
          name: "Masseuse",
          surname: "Test",
          email: "test@masseuse.org",
          subscribed: false,
          masseur: true
        },
        client: null,
        facility: { id: 1, name: "test" }
      }
    },
    {
      massage: {
        id: 2,
        date: new Date(1000),
        ending: new Date(2000),
        masseuse: {
          sub: "m-sub2",
          name: "Masseuse",
          surname: "Test2",
          email: "test2@masseuse.org",
          subscribed: true,
          masseur: true
        },
        client: { sub: "test" },
        facility: { id: 1, name: "test" }
      }
    }
  ];
  const testLocalizer = {
    messages: {
      massage: {
        date: "date",
        time: "time",
        event: "event"
      }
    }
  };
  const wrapper = shallow(<CalendarView events={testEvents} localizer={testLocalizer} onCancel={testFunction} />);
  const rows = wrapper.find(EventRow);

  expect(rows.get(0).props.event).toBe(testEvents[0]);
  expect(rows.get(1).props.event).toBe(testEvents[1]);
  expect(rows.get(0).props.onCancel).toBe(testFunction);
  expect(rows.get(1).props.onCancel).toBe(testFunction);
  expect(testFunction).not.toHaveBeenCalled();

  expect(wrapper).toMatchSnapshot();
});
