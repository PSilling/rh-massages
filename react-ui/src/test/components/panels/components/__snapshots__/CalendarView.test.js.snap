// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content with correct props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CalendarView
    events={
      Array [
        Object {
          "massage": Object {
            "client": null,
            "date": 1970-01-01T00:00:00.000Z,
            "ending": 1970-01-01T00:00:01.000Z,
            "facility": Object {
              "id": 1,
              "name": "test",
            },
            "id": 1,
            "masseuse": Object {
              "email": "test@masseuse.org",
              "masseur": true,
              "name": "Masseuse",
              "sub": "m-sub",
              "subscribed": false,
              "surname": "Test",
            },
          },
        },
        Object {
          "massage": Object {
            "client": Object {
              "sub": "test",
            },
            "date": 1970-01-01T00:00:01.000Z,
            "ending": 1970-01-01T00:00:02.000Z,
            "facility": Object {
              "id": 1,
              "name": "test",
            },
            "id": 2,
            "masseuse": Object {
              "email": "test2@masseuse.org",
              "masseur": true,
              "name": "Masseuse",
              "sub": "m-sub2",
              "subscribed": true,
              "surname": "Test2",
            },
          },
        },
      ]
    }
    localizer={
      Object {
        "messages": Object {
          "massage": Object {
            "date": "date",
            "event": "event",
            "time": "time",
          },
        },
      }
    }
    onCancel={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <thead>
          <tr>
            <th />
            <th />
            <th />
          </tr>
        </thead>,
        <tbody>
          <EventRow
            event={
              Object {
                "massage": Object {
                  "client": null,
                  "date": 1970-01-01T00:00:00.000Z,
                  "ending": 1970-01-01T00:00:01.000Z,
                  "facility": Object {
                    "id": 1,
                    "name": "test",
                  },
                  "id": 1,
                  "masseuse": Object {
                    "email": "test@masseuse.org",
                    "masseur": true,
                    "name": "Masseuse",
                    "sub": "m-sub",
                    "subscribed": false,
                    "surname": "Test",
                  },
                },
              }
            }
            onCancel={[MockFunction]}
          />
          <EventRow
            event={
              Object {
                "massage": Object {
                  "client": Object {
                    "sub": "test",
                  },
                  "date": 1970-01-01T00:00:01.000Z,
                  "ending": 1970-01-01T00:00:02.000Z,
                  "facility": Object {
                    "id": 1,
                    "name": "test",
                  },
                  "id": 2,
                  "masseuse": Object {
                    "email": "test2@masseuse.org",
                    "masseur": true,
                    "name": "Masseuse",
                    "sub": "m-sub2",
                    "subscribed": true,
                    "surname": "Test2",
                  },
                },
              }
            }
            onCancel={[MockFunction]}
          />
        </tbody>,
      ],
      "hover": true,
      "responsive": true,
      "responsiveTag": "div",
      "size": "sm",
      "striped": true,
      "tag": "table",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <th />
            <th />
            <th />
          </tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <th />,
              <th />,
              <th />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "th",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "th",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "th",
            },
          ],
          "type": "tr",
        },
        "type": "thead",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <EventRow
              event={
                Object {
                  "massage": Object {
                    "client": null,
                    "date": 1970-01-01T00:00:00.000Z,
                    "ending": 1970-01-01T00:00:01.000Z,
                    "facility": Object {
                      "id": 1,
                      "name": "test",
                    },
                    "id": 1,
                    "masseuse": Object {
                      "email": "test@masseuse.org",
                      "masseur": true,
                      "name": "Masseuse",
                      "sub": "m-sub",
                      "subscribed": false,
                      "surname": "Test",
                    },
                  },
                }
              }
              onCancel={[MockFunction]}
            />,
            <EventRow
              event={
                Object {
                  "massage": Object {
                    "client": Object {
                      "sub": "test",
                    },
                    "date": 1970-01-01T00:00:01.000Z,
                    "ending": 1970-01-01T00:00:02.000Z,
                    "facility": Object {
                      "id": 1,
                      "name": "test",
                    },
                    "id": 2,
                    "masseuse": Object {
                      "email": "test2@masseuse.org",
                      "masseur": true,
                      "name": "Masseuse",
                      "sub": "m-sub2",
                      "subscribed": true,
                      "surname": "Test2",
                    },
                  },
                }
              }
              onCancel={[MockFunction]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "event": Object {
                "massage": Object {
                  "client": null,
                  "date": 1970-01-01T00:00:00.000Z,
                  "ending": 1970-01-01T00:00:01.000Z,
                  "facility": Object {
                    "id": 1,
                    "name": "test",
                  },
                  "id": 1,
                  "masseuse": Object {
                    "email": "test@masseuse.org",
                    "masseur": true,
                    "name": "Masseuse",
                    "sub": "m-sub",
                    "subscribed": false,
                    "surname": "Test",
                  },
                },
              },
              "onCancel": [MockFunction],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "event": Object {
                "massage": Object {
                  "client": Object {
                    "sub": "test",
                  },
                  "date": 1970-01-01T00:00:01.000Z,
                  "ending": 1970-01-01T00:00:02.000Z,
                  "facility": Object {
                    "id": 1,
                    "name": "test",
                  },
                  "id": 2,
                  "masseuse": Object {
                    "email": "test2@masseuse.org",
                    "masseur": true,
                    "name": "Masseuse",
                    "sub": "m-sub2",
                    "subscribed": true,
                    "surname": "Test2",
                  },
                },
              },
              "onCancel": [MockFunction],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "tbody",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
              <th />
              <th />
              <th />
            </tr>
          </thead>,
          <tbody>
            <EventRow
              event={
                Object {
                  "massage": Object {
                    "client": null,
                    "date": 1970-01-01T00:00:00.000Z,
                    "ending": 1970-01-01T00:00:01.000Z,
                    "facility": Object {
                      "id": 1,
                      "name": "test",
                    },
                    "id": 1,
                    "masseuse": Object {
                      "email": "test@masseuse.org",
                      "masseur": true,
                      "name": "Masseuse",
                      "sub": "m-sub",
                      "subscribed": false,
                      "surname": "Test",
                    },
                  },
                }
              }
              onCancel={[MockFunction]}
            />
            <EventRow
              event={
                Object {
                  "massage": Object {
                    "client": Object {
                      "sub": "test",
                    },
                    "date": 1970-01-01T00:00:01.000Z,
                    "ending": 1970-01-01T00:00:02.000Z,
                    "facility": Object {
                      "id": 1,
                      "name": "test",
                    },
                    "id": 2,
                    "masseuse": Object {
                      "email": "test2@masseuse.org",
                      "masseur": true,
                      "name": "Masseuse",
                      "sub": "m-sub2",
                      "subscribed": true,
                      "surname": "Test2",
                    },
                  },
                }
              }
              onCancel={[MockFunction]}
            />
          </tbody>,
        ],
        "hover": true,
        "responsive": true,
        "responsiveTag": "div",
        "size": "sm",
        "striped": true,
        "tag": "table",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th />
              <th />
              <th />
            </tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <th />,
                <th />,
                <th />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": "th",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": "th",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": "th",
              },
            ],
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <EventRow
                event={
                  Object {
                    "massage": Object {
                      "client": null,
                      "date": 1970-01-01T00:00:00.000Z,
                      "ending": 1970-01-01T00:00:01.000Z,
                      "facility": Object {
                        "id": 1,
                        "name": "test",
                      },
                      "id": 1,
                      "masseuse": Object {
                        "email": "test@masseuse.org",
                        "masseur": true,
                        "name": "Masseuse",
                        "sub": "m-sub",
                        "subscribed": false,
                        "surname": "Test",
                      },
                    },
                  }
                }
                onCancel={[MockFunction]}
              />,
              <EventRow
                event={
                  Object {
                    "massage": Object {
                      "client": Object {
                        "sub": "test",
                      },
                      "date": 1970-01-01T00:00:01.000Z,
                      "ending": 1970-01-01T00:00:02.000Z,
                      "facility": Object {
                        "id": 1,
                        "name": "test",
                      },
                      "id": 2,
                      "masseuse": Object {
                        "email": "test2@masseuse.org",
                        "masseur": true,
                        "name": "Masseuse",
                        "sub": "m-sub2",
                        "subscribed": true,
                        "surname": "Test2",
                      },
                    },
                  }
                }
                onCancel={[MockFunction]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "event": Object {
                  "massage": Object {
                    "client": null,
                    "date": 1970-01-01T00:00:00.000Z,
                    "ending": 1970-01-01T00:00:01.000Z,
                    "facility": Object {
                      "id": 1,
                      "name": "test",
                    },
                    "id": 1,
                    "masseuse": Object {
                      "email": "test@masseuse.org",
                      "masseur": true,
                      "name": "Masseuse",
                      "sub": "m-sub",
                      "subscribed": false,
                      "surname": "Test",
                    },
                  },
                },
                "onCancel": [MockFunction],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "event": Object {
                  "massage": Object {
                    "client": Object {
                      "sub": "test",
                    },
                    "date": 1970-01-01T00:00:01.000Z,
                    "ending": 1970-01-01T00:00:02.000Z,
                    "facility": Object {
                      "id": 1,
                      "name": "test",
                    },
                    "id": 2,
                    "masseuse": Object {
                      "email": "test2@masseuse.org",
                      "masseur": true,
                      "name": "Masseuse",
                      "sub": "m-sub2",
                      "subscribed": true,
                      "surname": "Test2",
                    },
                  },
                },
                "onCancel": [MockFunction],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "tbody",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
